### PR TITLE
feat(apollo-parser, apollo-compiler): support parsing and validating FieldSet

### LINKED ISSUE #681: Add parsing + validation support for `FieldSet` scalars
Apollo Federation has multiple directives that take an argument of type [`FieldSet`](https://www.apollographql.com/docs/federation/subgraph-spec/#scalar-fieldset) to designate one or more fields of an object (or interface?) type. `FieldSet` is a custom scalar serialized as a string. Its value is expected to be parsed as a [`SelectionSet`](https://spec.graphql.org/October2021/#SelectionSet) with optional outer brackets. (This is GraphQL syntax inside of a string in GraphQL syntax.)

Currently apollo-rs can only parse GraphQL documents, but we’d rather Rust federation not need string concatenation hacks to create a parser input that looks like a document.

## API sketch

In apollo-parser:

* Give `apollo_parser::SyntaxTree` a `Cst` type parameter
* `apollo_parser::Parser::new(string).parse()` now returns `SyntaxTree<apollo_parser::cst::Document>` instead of just `SyntaxTree`
* Add `apollo_parser::Parser::new(string).parse_fieldset()` which returns `SyntaxTree<cst::SelectionSet>`
* The `SyntaxTree::document(self) -> cst::Document` method now only exists on `SyntaxTree<cst::Document>`
* Add `SyntaxTree::selection_set(self) -> cst::SelectionSet` for `SyntaxTree<cst::SelectionSet>`

In apollo-compiler:

* Add a top-level free function:
  ```rust
  fn parse_fieldset(
      schema: &Schema,
      type_name: impl Into<ast::NamedType>,
      source_text: impl Into<String>,
      path: impl AsRef<Path>,
  ) -> executable::SelectionSet
  ```
* Add a `Parser` method with the same signature

### PR DESCRIPTION
**Pull request by lrlna**
This starts work on supporting FieldSet.

The current items in scope:
- [x] support FieldSet parsing in apollo_parser
- [x] create `SyntaxTree::selection_set` getter method for accessing selection set created from FieldSet parsing
- [x] incorporate FieldSet parsing in apollo_compiler
- [x] create FieldSet validation in apollo_compiler that only runs specific validation rules

closes #681

### CODE REVIEW

File: crates/apollo-parser/src/parser/mod.rs [hunk 1]
```diff
@@ -131,13 +135,46 @@ impl<'a> Parser<'a> {
...
+
+        match builder {
+            syntax_tree::SyntaxTreeWrapper::Document(tree) => tree,
+            syntax_tree::SyntaxTreeWrapper::FieldSet(_) => {
+                unreachable!("parse constructor can only construct a document")
+            }
+        }
+    }
+
+    pub fn parse_fieldset(mut self) -> SyntaxTree<SelectionSet> {
```
Comment (lrlna) [inline crates/apollo-parser/src/parser/mod.rs:159]: i am almost certain i'd like this to be called `parse_selection_set`. `fieldset` is a federation specific term based on a non-spec scalar definition. I'd rather have the parser be by the spec

File: crates/apollo-compiler/CHANGELOG.md [hunk 2]
```diff
@@ -34,7 +34,13 @@ Assorted `Schema` API changes by [SimonSapin] in [pull/678]:
 
 ## Features
 
-- Add opt-in configuration for “orphan” extensions to be “adopted”, by [SimonSapin] in [pull/678]
+- **Add `executable::FieldSet` for a selection set with optional outer brackets - [lrlna], [pull/685] fixing [issue/681]**
+  This is intended to parse string value of a [`FieldSet` custom scalar][fieldset]
+  used in some Apollo Federation directives, in the context of a specific schema and type.
```
Comment (lrlna) [inline crates/apollo-compiler/CHANGELOG.md:39]:
```suggestion
  used in some Apollo Federation directives in the context of a specific schema and type.
```

File: crates/apollo-compiler/CHANGELOG.md [hunk 3]
```diff
@@ -34,7 +34,13 @@ Assorted `Schema` API changes by [SimonSapin] in [pull/678]:
...
+  Its `validate` method calls the subset of validation rules relevan to a selection set
```
Comment (lrlna) [inline crates/apollo-compiler/CHANGELOG.md:40]:
```suggestion
  Its `validate` method calls a subset of validation rules relevant to selection sets.
```

File: crates/apollo-parser/CHANGELOG.md [hunk 4]
```diff
@@ -17,6 +17,21 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## Documentation -->
 
+# [x.x.x] (unreleased) - 2022-mm-dd
+
+## Features
+- **`parse_field_set` parses a selection set with optional outer brackets - [lrlna], [pull/685] fixing [issue/681]**
+  This returns a `SyntaxTree<SelectionSet>` which instead of `.document() -> cst::Document`
+  has `.field_set() -> cst::SelectionSet`.
+  This is intended to the parse string value of a [`FieldSet` custom scalar][fieldset]
```
Comment (lrlna) [inline crates/apollo-parser/CHANGELOG.md:26]:
```suggestion
  This is intended to parse string value of a [`FieldSet` custom scalar][fieldset]
```

**Review by SimonSapin — APPROVED** [review]

**Pull request merged**